### PR TITLE
Add Moonpay to Add Funds

### PR DESCRIPTION
**Pull request by erdemyerebasmaz**
This PR implements [BU-178](https://breeztech.atlassian.net/browse/BU-178) & [BU-180](https://breeztech.atlassian.net/browse/BU-180).

An [intermediary screen](https://zpl.io/VKykvAq) is implemented to reduce clutter in Add Funds screen.
Users are now able to buy bitcoins using MoonPay vendor. Clicking Buy Bitcoin button routes users to moonpay's website using a Vebview, provided with various parameters to ease the process.
Such as currencyCode, walletAddress, maxQuoteCurrencyAmount etc.

Once the order is complete, it's information is passed to AddFundsBloc. If the address associated with the transaction is in unConfirmedAddresses "Your MoonPay order is being processed." message is shown. Which will expire once the transaction is confirmed or after a set amount of time has passed.

TODO:
 - Determine the expiration time,
 - Set apiKey to publishable key and test on production,
 - ~Determine which error messages to be removed from deposit_to_btc_address_page~

### CODE REVIEW

File: lib/bloc/account/account_model.dart [hunk 1]
```diff
@@ -20,34 +20,39 @@ enum SyncUIState {
...
-            showConnectProgress: json["showConnectProgress"] ?? false);
+            showConnectProgress: json["showConnectProgress"] ?? false,
+            moonpayIpCheck: json["moonpayIpCheck"] ?? true,
+            failePaymentBehavior: BugReportBehavior.values[json["failePaymentBehavior"] ?? 0]);
 
   Map<String, dynamic> toJson() {
     return {
       "ignoreWalletBalance": ignoreWalletBalance,
+      "showConnectProgress": showConnectProgress ?? false,
+      "moonpayIpCheck": moonpayIpCheck ?? true,
```
Comment (roeierez) [inline lib/bloc/account/account_model.dart]: We don't need the null fallback here for "`moonpayIpCheck`" and "`showConnectProgress`". They will always have value.

File: lib/bloc/account/add_funds_bloc.dart [hunk 2]
```diff
@@ -11,19 +14,65 @@ class AddFundsBloc {
   final _addFundResponseController = new StreamController<AddFundResponse>();
   Stream<AddFundResponse> get addFundResponseStream => _addFundResponseController.stream;
 
+  final _moonPayOrderController = new StreamController<MoonpayOrder>();
+  Sink<MoonpayOrder> get moonPayOrderSink => _moonPayOrderController.sink;
```
Comment (roeierez) [inline lib/bloc/account/add_funds_bloc.dart]: Why do we need additional MoonpayOrder controller (sink and stream)?
Can't we use the `_orderController ` also for pending orders?

File: lib/bloc/account/add_funds_bloc.dart [hunk 3]
```diff
@@ -11,19 +14,65 @@ class AddFundsBloc {
...
           .then((reply) => _addFundResponseController.add(new AddFundResponse(reply)))
           .catchError(_addFundResponseController.addError);
       })
       .onDone(_dispose);
+
+    _initializeWithMoonpayOrder(injector);
+    _listenToMoonpayOrder(injector);
+  }
+
+  _initializeWithMoonpayOrder(ServiceInjector injector) {
```
Comment (roeierez) [inline lib/bloc/account/add_funds_bloc.dart]: I would simplify this function by always adding the last `MoonpayOrder` to the stream and no need to delete it from the storage as well.
Something like:
```suggestion
  var order = preferences.getString("pendingMoonpayOrderAddress");
  if (order != null) {
      _moonPayOrderController.add(MoonpayOrder(
            preferences.getString("pendingMoonpayOrderAddress") ?? null, 
            preferences.getInt("pendingMoonpayOrderTimestamp") ?? null))
   }
```

The UI only needs to filter those expired (already does that).

File: lib/bloc/account/add_funds_bloc.dart [hunk 4]
```diff
@@ -11,19 +14,65 @@ class AddFundsBloc {
...
+    _orderController.stream.listen((order) async {
+      injector.sharedPreferences.then((preferences) {
+        _saveMoonpayOrder(preferences, order);
+        _moonPayOrderController.add(order);
+      });
+    });
+  }
+
+  _saveMoonpayOrder(SharedPreferences preferences, MoonpayOrder order) {
+    preferences.setString("pendingMoonpayOrderAddress", order.address);
```
Comment (roeierez) [inline lib/bloc/account/add_funds_bloc.dart]: I think it is better that MoonpayOrder class will handle its own json serialization and we can write its value under one key.

File: lib/routes/user/add_funds/add_funds_page.dart [hunk 5]
```diff
@@ -43,71 +56,106 @@ class AddFundsState extends State<AddFundsPage> {
         _accountSubscription.cancel();
       }
     });
+    _settingsSubscription = widget._accountBloc.accountSettingsStream.listen((settings) {
+      isIpAllowed(settings.moonpayIpCheck);
+    });
+    _moonPaySubscription = _addFundsBloc.moonPayOrderStream.listen((order) {
```
Comment (roeierez) [inline lib/routes/user/add_funds/add_funds_page.dart]: I don't think we need a subscription here, just use the stream in the build method.

File: lib/routes/user/add_funds/add_funds_page.dart [hunk 6]
```diff
@@ -43,71 +56,106 @@ class AddFundsState extends State<AddFundsPage> {
...
+      setState(() {
+        _moonPayOrder = order;
+      });
+    });
+    _checkMoonpayOrderExpiration();
+  }
+
+  isIpAllowed(bool ipCheckSetting) async {
```
Comment (roeierez) [inline lib/routes/user/add_funds/add_funds_page.dart]: all ip checking and the decision if to show moonpay or not should be in the bloc.
Also the moonpay URL should be generated in the bloc.
I think we can add moonpayURL in AddFundsResponse model and show the moonpay menu only if this field has value.
This way we can encapsulate the ip logic in the bloc without making the UI complex.

File: lib/routes/user/add_funds/add_funds_page.dart [hunk 7]
```diff
@@ -120,121 +168,195 @@ class AddFundsState extends State<AddFundsPage> {
...
-          text: hasError ? "RETRY" : "CLOSE",
-          onPressed: () {
-            if (hasError) {
-              _addFundsBloc.addFundRequestSink.add(null);
-            } else {
-              Navigator.of(context).pop();
-            }
-          });
+  Widget _buildMoonpayButton(AddFundResponse response) {
+    String baseUrl = "https://buy-staging.moonpay.io";
```
Comment (roeierez) [inline lib/routes/user/add_funds/add_funds_page.dart]: The moonpay url should be generated in the bloc. Also the `apiKey` should be fetched from a config file.

File: lib/routes/user/add_funds/add_funds_page.dart [hunk 8]
```diff
@@ -120,121 +168,195 @@ class AddFundsState extends State<AddFundsPage> {
...
+    String apiKey = "pk_test_AZskxvTXb0rpsI7o2GCdmzs8jeST9d";
+    String currencyCode = "btc";
+    String walletAddress = "n4VQ5YdHf7hLQ2gWQYYrcxoE5B7nWuDFNF";
```
Comment (roeierez) [inline lib/routes/user/add_funds/add_funds_page.dart]: Hard coded address...

File: lib/routes/user/marketplace/vendor_webview.dart [hunk 9]
```diff
@@ -62,53 +64,85 @@ class VendorWebViewPageState extends State<VendorWebViewPage> {
...
                 });
               }
-            });            
-        }
-      });
+            });
+          }
+        });
+      } else {
+        String loadedURL;
```
Comment (roeierez) [inline lib/routes/user/marketplace/vendor_webview.dart]: I don't think the internals of moonpay logic belongs here as they are not related to VendorWebView.
I also see that most of the logic in this class is not relevant to moonpay as well so I think it is better to create a separate widget "MoonpayWebview" with the specific logic.

File: lib/routes/user/marketplace/vendor_webview.dart [hunk 10]
```diff
@@ -62,53 +64,85 @@ class VendorWebViewPageState extends State<VendorWebViewPage> {
...
+        _widgetWebview.onStateChanged.listen((state) async {
+          if (state.type == WebViewState.finishLoad && loadedURL != state.url) {
+            loadedURL = state.url;
+            var script = "function getParameterByName(name) {    \n" +
```
Comment (roeierez) [inline lib/routes/user/marketplace/vendor_webview.dart]: The script should be in a resource file.

File: lib/routes/user/add_funds/add_funds_page.dart (hunk 8 above)
Comment (erdemyerebasmaz) [reply lib/routes/user/add_funds/add_funds_page.dart]: This is hardcoded for test mode, will be changed to current address when we switch to production.

File: lib/routes/user/marketplace/vendor_webview.dart (hunk 9 above)
Comment (erdemyerebasmaz) [reply lib/routes/user/marketplace/vendor_webview.dart]: [Fixed](https://github.com/breez/breezmobile/pull/162/commits/ec80a3164e05ea547c90c5d39dd519e0563b431e)

File: lib/bloc/account/account_model.dart (hunk 1 above)
Comment (erdemyerebasmaz) [reply lib/bloc/account/account_model.dart]: [Fixed](https://github.com/breez/breezmobile/pull/162/commits/9bb1eabe77c2447ccb86553ccd0231efbb76737f)

File: lib/routes/user/marketplace/vendor_webview.dart (hunk 10 above)
Comment (erdemyerebasmaz) [reply lib/routes/user/marketplace/vendor_webview.dart]: [Fixed](https://github.com/breez/breezmobile/pull/162/commits/52282b8d816a7acb283a541d8d78a892bdb550ff)

File: lib/routes/user/add_funds/add_funds_page.dart (hunk 7 above)
Comment (erdemyerebasmaz) [reply lib/routes/user/add_funds/add_funds_page.dart]: Fixed
[1](https://github.com/breez/breezmobile/pull/162/commits/e2cbce256ee9aeba8b56b024e9eb6c650dbbb913)
[2](https://github.com/breez/breezmobile/pull/162/commits/2c95697004ad429b7857223898348aa9ab1ec3c0)

File: lib/bloc/account/add_funds_bloc.dart (hunk 3 above)
Comment (erdemyerebasmaz) [reply lib/bloc/account/add_funds_bloc.dart]: [Fixed](https://github.com/breez/breezmobile/pull/162/commits/96486619e86debc7d6514689a54adbb2e76a63bd)

File: lib/bloc/account/add_funds_bloc.dart (hunk 4 above)
Comment (erdemyerebasmaz) [reply lib/bloc/account/add_funds_bloc.dart]: [Fixed](https://github.com/breez/breezmobile/pull/162/commits/8f9ff8cb02826657a80bf08ddb67b2dd7453a0f4)

File: lib/bloc/account/add_funds_bloc.dart (hunk 2 above)
Comment (erdemyerebasmaz) [reply lib/bloc/account/add_funds_bloc.dart]: It's removed in [9648661](https://github.com/breez/breezmobile/pull/162/commits/96486619e86debc7d6514689a54adbb2e76a63bd)

File: lib/routes/user/add_funds/add_funds_page.dart (hunk 5 above)
Comment (erdemyerebasmaz) [reply lib/routes/user/add_funds/add_funds_page.dart]: [Removed in](https://github.com/breez/breezmobile/pull/162/commits/e575a05e69ee9d79c65ba65bc54d6f1b73380a1c)

File: lib/routes/user/add_funds/add_funds_page.dart (hunk 6 above)
Comment (erdemyerebasmaz) [reply lib/routes/user/add_funds/add_funds_page.dart]: [Fixed](https://github.com/breez/breezmobile/pull/162/commits/e575a05e69ee9d79c65ba65bc54d6f1b73380a1c)

File: lib/bloc/account/add_funds_bloc.dart [hunk 11]
```diff
@@ -1,29 +1,107 @@
...
+      Map<String, dynamic> settings = json.decode(accountSettings);
+      if (settings["moonpayIpCheck"]) {
+        _isIpAllowed().then((ipIsAllowed) {
+          if (ipIsAllowed) _createMoonpayUrl();
+        });
+      }
+    }
+  }
+
+  Future<bool> _isIpAllowed() async {
```
Comment (roeierez) [inline lib/bloc/account/add_funds_bloc.dart]: NIT: rename to `_isIPMoonpayAllowed` ?

File: lib/bloc/account/add_funds_bloc.dart [hunk 12]
```diff
@@ -1,29 +1,107 @@
...
+    }).onDone(_dispose);
+    _listenAccountSettings(injector);
+    _handleMoonpayOrders(injector);
+  }
+
+  String get moonPayURL => _moonPayURL;
+
+  _listenAccountSettings(ServiceInjector injector) async {
+    var preferences = await injector.sharedPreferences;
+    var accountSettings = preferences.getString(ACCOUNT_SETTINGS_PREFERENCES_KEY);
```
Comment (roeierez) [inline lib/bloc/account/add_funds_bloc.dart]: Ideally this value should be saved at the same place it is read from.
I would move the save action from the AccountBloc to the AddFundsBloc.

File: lib/routes/user/add_funds/add_funds_page.dart [hunk 13]
```diff
@@ -43,71 +51,89 @@ class AddFundsState extends State<AddFundsPage> {
...
-    } else if (unconfirmedTxID?.isNotEmpty == true ||
-        account.processingWithdrawal) {
+      errorMessage = 'You\'d be able to add funds after Breez is finished bootstrapping.';
+    } else if (unconfirmedTxID?.isNotEmpty == true || account.processingWithdrawal) {
       errorMessage =
           'Breez is processing your previous ${waitingDepositConfirmation || account.processingBreezConnection ? "deposit" : "withdrawal"}. You will be able to add more funds once this operation is completed.';
     } else if (response != null && response.errorMessage.isNotEmpty) {
       errorMessage = response.errorMessage;
+    } else if (_moonPayOrder?.address != null &&
+        account?.addedFundsReply?.unConfirmedAddresses
```
Comment (roeierez) [inline lib/routes/user/add_funds/add_funds_page.dart]: Although the `addedFundsReply` is public field it wasn't the intention as it is a protobuf structure.
The access should be via the `SwapFundStatus` that wraps it.
I would add a getter there for `unconfirmedAddresses` same as the `refundableAddresses` that exists there.

File: lib/routes/user/add_funds/add_funds_page.dart [hunk 14]
```diff
@@ -120,121 +146,185 @@ class AddFundsState extends State<AddFundsPage> {
...
-          onPressed: () {
-            if (hasError) {
-              _addFundsBloc.addFundRequestSink.add(null);
-            } else {
-              Navigator.of(context).pop();
-            }
-          });
+  Widget _buildMoonpayButton(AddFundResponse response) {
+    String walletAddress = "n4VQ5YdHf7hLQ2gWQYYrcxoE5B7nWuDFNF"; // Will switch to response?.address when we use public apiKey
+    String moonPayURL = _addFundsBloc.moonPayURL + "&walletAddress=$walletAddress";
```
Comment (roeierez) [inline lib/routes/user/add_funds/add_funds_page.dart]: It is not a practice in our blocs to expose public getter. Instead the blocs provide the data via their stream interface.

File: lib/bloc/account/add_funds_bloc.dart (hunk 11 above)
Comment (erdemyerebasmaz) [reply lib/bloc/account/add_funds_bloc.dart]: [Renamed to _isIPMoonpayAllowed](https://github.com/breez/breezmobile/pull/162/commits/55c902c244bd6e62994f90a9df275198276d0b73)

File: lib/routes/user/add_funds/add_funds_page.dart (hunk 14 above)
Comment (erdemyerebasmaz) [reply lib/routes/user/add_funds/add_funds_page.dart]: https://github.com/breez/breezmobile/pull/162/commits/71dc2301a1f97ae791929615644062c444d2ab5b

File: lib/routes/user/add_funds/add_funds_page.dart (hunk 13 above)
Comment (erdemyerebasmaz) [reply lib/routes/user/add_funds/add_funds_page.dart]: https://github.com/breez/breezmobile/pull/162/commits/eb57e859db19800afe4054b9d63ef1bcb0617671

**Pull request merged**